### PR TITLE
Make allocateNode run async in case of block offline event

### PR DESCRIPTION
**Pull request by congqixia**
Related issue: #17099 

/kind bug

Signed-off-by: Congqi Xia <congqi.xia@zilliz.com>

### CODE REVIEW

File: internal/querycoord/query_coord.go [hunk 1]
```diff
@@ -471,9 +434,11 @@ func (qc *QueryCoord) handleNodeEvent(ctx context.Context) {
...
 				}
-				if err := qc.allocateNode(serverID); err != nil {
-					log.Error("unable to allcoate node", zap.Int64("nodeID", serverID), zap.Error(err))
-				}
+				go func(serverID int64) {
+					if err := qc.allocateNode(serverID); err != nil {
+						log.Error("unable to allcoate node", zap.Int64("nodeID", serverID), zap.Error(err))
+					}
+				}(serverID)
 				qc.metricsCacheManager.InvalidateSystemInfoMetrics()
```
Comment (XuanYang-cn) [inline internal/querycoord/query_coord.go:442]: I'm not sure what exactly L442 does, but this change makes the executing order uncertain, which might be a problem?

Also I'm curious about why `registerNode` failure doesn't need to call L442, and why L442 needs to be called over and over again.

Comment (congqixia) [reply internal/querycoord/query_coord.go:442]: It's ok to make new querynode register not randomly.

L442 is related to the metrics cache, if the new node registered successfully, the online node list shall be changed so the cached metrics shall be invalidated.

File: internal/querycoord/meta.go [hunk 2]
```diff
@@ -1282,6 +1283,11 @@ func (m *MetaReplica) getReplicasByCollectionID(collectionID int64) ([]*milvuspb
 	return replicas, nil
 }
 
+// applyReplicaBalancePlan applies replica balance plan to replica info.
+func (m *MetaReplica) applyReplicaBalancePlan(p *balancePlan) error {
+	return m.replicas.ApplyBalancePlan(p)
```
Comment (yah01) [inline internal/querycoord/meta.go:1288]: need to write the modified replicas to etcd

File: internal/querycoord/replica.go [hunk 3]
```diff
@@ -112,3 +122,102 @@ func (rep *ReplicaInfos) Remove(replicaIds ...UniqueID) {
...
+		replicasChanged = append(replicasChanged, sourceReplica)
+	}
+	if targetReplica != nil {
+		// add node to replica
+		targetReplica.NodeIds = append(targetReplica.NodeIds, p.nodeID)
+		replicasChanged = append(replicasChanged, targetReplica)
+	}
+
+	// save to etcd first
+	if len(replicasChanged) > 0 {
```
Comment (yah01) [inline internal/querycoord/replica.go:165]: it can be simple to implement these in MetaReplica, which allows you to use the setReplica() method.

Comment (yah01) [reply internal/querycoord/replica.go:165]: This [example](https://github.com/milvus-io/milvus/pull/17218/files#diff-ade465f7861613623d78a92764a878a2fcf296fccc1f260a64f3829a2b3b006dR1303) may help

Comment (congqixia) [reply internal/querycoord/replica.go:165]: @yah01 tried this way but cannot keep this whole operation atomic.

File: internal/querycoord/replica.go [hunk 4]
```diff
@@ -112,3 +122,102 @@ func (rep *ReplicaInfos) Remove(replicaIds ...UniqueID) {
...
+			key := fmt.Sprintf("%s/%d", ReplicaMetaPrefix, info.ReplicaID)
+			data[key] = string(infoBytes)
+		}
+		err := kv.MultiSave(data)
+		if err != nil {
+			return err
+		}
+	}
+
+	// apply change to in-memory meta
```
Comment (yah01) [inline internal/querycoord/replica.go:183]: we should provide a private insert() method to handle this, ref Insert()

File: internal/querycoord/replica.go [hunk 5]
```diff
@@ -112,3 +122,102 @@ func (rep *ReplicaInfos) Remove(replicaIds ...UniqueID) {
...
+
+	var sourceReplica, targetReplica *milvuspb.ReplicaInfo
+
+	// check source and target replica ids are valid
+	if p.sourceReplica != invalidReplicaID {
+		replica, ok := rep.replicas[p.sourceReplica]
+		if !ok {
+			return errors.New("replica not found")
+		}
+		sourceReplica = proto.Clone(replica).(*milvuspb.ReplicaInfo)
```
Comment (yah01) [inline internal/querycoord/replica.go:139]: This code occurs frequently, we need a private get() method just like Get().

File: internal/querycoord/replica.go [hunk 6]
```diff
@@ -112,3 +122,102 @@ func (rep *ReplicaInfos) Remove(replicaIds ...UniqueID) {
...
+	if len(replicasChanged) > 0 {
+		data := make(map[string]string)
+
+		for _, info := range replicasChanged {
+			infoBytes, err := proto.Marshal(info)
+			if err != nil {
+				return err
+			}
+
+			key := fmt.Sprintf("%s/%d", ReplicaMetaPrefix, info.ReplicaID)
```
Comment (yah01) [inline internal/querycoord/replica.go:174]: Should Meta provide a method to save replica? Or just let insert() method do this

File: internal/querycoord/replica.go (hunk 5 above)
Comment (congqixia) [reply internal/querycoord/replica.go]: fixed

File: internal/querycoord/replica.go (hunk 6 above)
Comment (congqixia) [reply internal/querycoord/replica.go:191]: This save operation may contain multiple replica info entries, temporary use dedicated code. We could refactor that if this situation changes.

**Pull request merged**